packages/migrations/src/actions/2023.01.18T11.03.41.registry-v2.sql

```diff
@@ -0,0 +1,117 @@
+-- 
+
+-- Describes an action that was performed on a registry.
+-- PUSH - a new schema was pushed to the registry, an existing schema was updated
+-- DELETE - a schema was deleted from the registry
+CREATE TYPE schema_registry_action AS ENUM ('PUSH', 'DELETE');
+
+-- Holds the actions that were performed on the registry.
+-- Every time a schema is pushed or deleted, a new entry is created.
+CREATE TABLE public.schema_log (
+  id uuid PRIMARY KEY DEFAULT uuid_generate_v4(),
+  author text NOT NULL,
+  created_at timestamp with time zone NOT NULL DEFAULT NOW(),
+  service_name text,
+  service_url text,
+  sdl text,
+  metadata text,
+  commit text NOT NULL,
+  action schema_registry_action NOT NULL,
+  target_id uuid NOT NULL REFERENCES public.targets(id) ON DELETE CASCADE,
+  project_id uuid NOT NULL REFERENCES public.projects(id) ON DELETE CASCADE
+);
+
+-- Describes the state of a schema in the registry.
+-- Groups together all the actions that were performed on a schema.
+-- The latest version is the current state of the schema.
+-- The last version that is composable is the current state of the schema available to the gateways.
+-- Every time a schema is pushed or deleted, a new version is created, that holds a reference to the actions performed in the previous version.
+CREATE TABLE public.schema_versions (
+  id uuid PRIMARY KEY DEFAULT uuid_generate_v4(),
+  created_at timestamp with time zone NOT NULL DEFAULT NOW(),
+  is_composable boolean NOT NULL,
+  base_schema text,
+  target_id uuid NOT NULL REFERENCES public.targets(id) ON DELETE CASCADE,
+  action_id uuid NOT NULL REFERENCES public.schema_log(id) ON DELETE CASCADE
+);
+
+-- Describes the relationship between a schema version and the actions that were performed previously but are related to the matching schema version.
+CREATE TABLE public.schema_version_to_log (
+  version_id uuid NOT NULL REFERENCES public.schema_versions(id) ON DELETE CASCADE,
+  action_id uuid NOT NULL REFERENCES public.schema_log(id) ON DELETE CASCADE,
+  PRIMARY KEY(version_id, action_id)
+);
+
+-- This is used to determine if the project is using the new registry model or the old one.
+ALTER TABLE public.projects
+  ADD COLUMN legacy_registry_model boolean NOT NULL DEFAULT FALSE;
```

Comment:
adding a column with a default value to an existing table will cause every row in the table to be rewritten, which will immediatly double the tables used space and also increase the migration runtime.

It might be better to add the column without a default value and do the default value mapping / interpret null as "false" within the business logic of the application.

Reply:
I don't think it will cause any perf issues